File: cache/cache-annotation-processor/src/main/java/ru/tinkoff/kora/cache/annotation/processor/CacheOperation.java
package ru.tinkoff.kora.cache.annotation.processor;

import com.squareup.javapoet.CodeBlock;

import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import java.util.List;

public record CacheOperation(Type type, List<CacheExecution> executions, Origin origin) {

    public record CacheExecution(String field, TypeElement type, DeclaredType superType, Contract contract, CacheKey cacheKey) {

        public enum Contract {
            SYNC,
            ASYNC
        }
    }

    public record CacheKey(DeclaredType type, CodeBlock code) {}

    public record Origin(String className, String methodName) {

        @Override
        public String toString() {
            return "[class=" + className + ", method=" + methodName + ']';
        }
    }

    public enum Type {
        GET,
        PUT,
        EVICT,
        EVICT_ALL
    }
}

File: cache/cache-symbol-processor/src/main/kotlin/ru/tinkoff/kora/cache/symbol/processor/CacheOperation.kt
package ru.tinkoff.kora.cache.symbol.processor

import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.symbol.KSType
import com.google.devtools.ksp.symbol.KSTypeArgument
import com.google.devtools.ksp.symbol.KSTypeReference
import com.squareup.kotlinpoet.CodeBlock

@KspExperimental
data class CacheOperation(
    val type: Type,
    val executions: List<CacheExecution>,
    val origin: Origin
) {

    data class CacheExecution(
        val field: String,
        val type: KSType,
        val superType: KSTypeReference,
        val cacheKey: CacheKey?
    )

    data class CacheKey(val type: KSTypeArgument, val code: CodeBlock)

    data class Origin(val className: String, val methodName: String) {
        override fun toString(): String = "[class=$className, method=$methodName]"
    }

    enum class Type {
        GET,
        PUT,
        EVICT,
        EVICT_ALL
    }
}
